java-squid/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValueRelation.java

```diff
@@ -64,46 +83,57 @@ private boolean equalsRelation(SymbolicValueRelation relation) {
   }
 
   private int[] key() {
-    if (v1.id() < v2.id()) {
-      return new int[] {v1.id(), v2.id()};
+    if (leftOp.id() < rightOp.id()) {
+      return new int[] {leftOp.id(), rightOp.id()};
     } else {
-      return new int[] {v2.id(), v1.id()};
+      return new int[] {rightOp.id(), leftOp.id()};
     }
   }
 
   @Override
   public String toString() {
     StringBuilder buffer = new StringBuilder();
-    buffer.append(v1);
-    buffer.append(' ');
+    buffer.append(leftOp);
     buffer.append(getOperand());
-    buffer.append(' ');
-    buffer.append(v2);
+    buffer.append(rightOp);
     return buffer.toString();
   }
 
+  @Override
   @CheckForNull
-  public Boolean impliedBy(Collection<SymbolicValueRelation> knownRelations) {
-    return impliedBy(knownRelations, new HashSet<SymbolicValueRelation>());
+  protected ProgramState checkRelation(BooleanConstraint booleanConstraint, ProgramState programState) {
+    RelationState relationState = resolveState(programState.getKnownRelations());
+    if (relationState.rejects(booleanConstraint)) {
+      return null;
+    }
+    return programState;
+  }
+
+  protected RelationState resolveState(Collection<SymbolicValueRelation> knownRelations) {
+    return resolveState(knownRelations, new HashSet<SymbolicValueRelation>());
   }
 
   @CheckForNull
-  public Boolean impliedBy(Collection<SymbolicValueRelation> knownRelations, Set<SymbolicValueRelation> usedRelations) {
+  protected RelationState resolveState(Collection<SymbolicValueRelation> knownRelations, Set<SymbolicValueRelation> usedRelations) {
     if (knownRelations.isEmpty()) {
-      return null;
+      return RelationState.UNDETERMINED;
     }
     if (usedRelations.size() > 200) {
```

Comment:
Is this covered by unit tests ? I am wondering if there is any possible circular substitution possible.